checker miss leads to cgen error when using sum type in var arg as array in static method

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.12, press to see full `v doctor` output</summary>

|V full version      |V 0.4.12 9772649250016a8558fe9bf07ba94f2c812fa9ef
|:-------------------|:-------------------
|OS                  |linux, "EndeavourOS Linux"
|Processor           |16 cpus, 64bit, little endian, AMD Ryzen 7 8845HS w/ Radeon 780M Graphics
|Memory              |16.11GB/30.64GB
|                    |
|V executable        |/home/lmp/Projects/v/v
|V last modified time|2025-09-30 07:15:01
|                    |
|V home dir          |OK, value: /home/lmp/Projects/v
|VMODULES            |OK, value: /home/lmp/.vmodules
|VTMP                |OK, value: /dev/shm
|Current working dir |OK, value: /home/lmp/Projects/v
|                    |
|Git version         |git version 2.51.0
|V git status        |weekly.2025.40-2-g97726492
|.git/config present |true
|                    |
|cc version          |cc (GCC) 15.2.1 20250813
|gcc version         |gcc (GCC) 15.2.1 20250813
|clang version       |clang version 20.1.8
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |shared:INFO: old sanity: 4.0.12-git|/opt/emscripten-llvm/bin
|glibc version       |ldd (GNU libc) 2.42

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  bugs/t.v && /home/lmp/Projects/v/bugs/t`
```v
module main

type Nil = u8

const c_nil = Nil(`\0`)

pub type Value = Nil | int | string | []Value | Map

pub fn (v Value) string() string {
	return '${v.str()}'
}

pub struct Map {
mut:
	m map[string]Value
}

pub fn Map.create(values ...Value) Map {
	assert values.len > 1
	assert values.len % 2 == 0
	mut m := Map{}
	for i := 0; i < values.len; i += 2 {
		m.set(values[i], values[i + 1])
	}
	return m
}

pub fn (mut m Map) set(key Value, value Value) {
	m.m[key.string()] = value
}

pub fn build_map(index_from int, index_to int) Map {
	mut kv_pairs := []Value{len: index_to - index_from, init: Value(c_nil)}
	for i := index_from; i < index_to; i++ {
		kv_pairs[i - index_from] = Value(i)
	}
	assert kv_pairs.len % 2 == 0
	// NOTE: this alternative works:
	// mut m := Map{}
	// for i := 0; i < kv_pairs.len; i += 2 {
	// 	m.set(kv_pairs[i], kv_pairs[i + 1])
	// }
	// return m
	// NOTE: this fails
	return Map.create(kv_pairs)
}

fn main() {
	m := build_map(2, 4)
	dump(m)
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: /dev/shm/t.01K6DDH13DJ9Z0CYK7DB8C337T.tmp.c:6645: error: cannot convert 'struct main__Value' to 'struct array'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

Something else than the C error, probably the `dump(m)`


> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1528,20 +1528,39 @@ fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool) ast.
 				c.expr(mut call_arg.expr)
 			}
 			if i == args_len - 1 {
-				if c.table.sym(typ).kind == .array && call_arg.expr !is ast.ArrayDecompose
-					&& c.table.sym(expected_type).kind !in [.sum_type, .interface]
-					&& !param.typ.has_flag(.generic) && expected_type != typ {
-					styp := c.table.type_to_str(typ)
-					elem_styp := c.table.type_to_str(expected_type)
+				kind := c.table.sym(typ).kind
+				is_decompose := call_arg.expr is ast.ArrayDecompose
+				mut allow_variadic_pass := false
+				if call_arg.expr is ast.Ident && !func.is_method {
+					ident := call_arg.expr as ast.Ident
+					if ident.obj is ast.Var {
+						var_obj := ident.obj as ast.Var
+						if var_obj.is_arg && c.table.cur_fn != unsafe { nil }
+							&& c.table.cur_fn.is_variadic && c.table.cur_fn.params.len > 0
+							&& ident.name == c.table.cur_fn.params.last().name && func.is_variadic {
+							cur_fn_param_sym := c.table.sym(c.table.cur_fn.params.last().typ)
+							if cur_fn_param_sym.info is ast.Array {
+								allow_variadic_pass = cur_fn_param_sym.info.elem_type == expected_type
+							}
+						}
+					}
+				}
+				styp := c.table.type_to_str(typ)
+				elem_styp := c.table.type_to_str(expected_type)
+				is_single_array_arg := args_len == 1 && i == 0
+				sum_type_needs_spread := c.table.sym(expected_type).kind == .sum_type
+					&& is_single_array_arg
+				if kind == .array && !is_decompose && !allow_variadic_pass
+					&& (c.table.sym(expected_type).kind !in [.sum_type, .interface]
+					|| sum_type_needs_spread) && !param.typ.has_flag(.generic)
+					&& expected_type != typ {
 					c.error('to pass `${call_arg.expr}` (${styp}) to `${func.name}` (which accepts type `...${elem_styp}`), use `...${call_arg.expr}`',
 						node.pos)
-				} else if call_arg.expr is ast.ArrayDecompose
-					&& c.table.sym(expected_type).kind == .sum_type
-					&& expected_type.idx() != typ.idx() {
-					expected_type_str := c.table.type_to_str(expected_type)
-					got_type_str := c.table.type_to_str(typ)
-					c.error('cannot use `...${got_type_str}` as `...${expected_type_str}` in argument ${
-						i + 1} to `${fn_name}`', call_arg.pos)
+				} else if is_decompose && !param.typ.has_flag(.generic)
+					&& c.table.sym(expected_type).kind != .interface
+					&& expected_type.idx() != typ.idx() && typ != ast.void_type {
+					c.error('cannot use `...${styp}` as `...${elem_styp}` in argument ${i + 1} to `${fn_name}`',
+						call_arg.pos)
 				}
 			}
 		} else {
@@ -2554,20 +2573,24 @@ fn (mut c Checker) method_call(mut node ast.CallExpr, mut continue_check &bool) 
 			}
 			typ := c.expr(mut arg.expr)
 			if i == node.args.len - 1 {
-				if c.table.sym(typ).kind == .array && arg.expr !is ast.ArrayDecompose
-					&& c.table.sym(expected_type).kind !in [.sum_type, .interface]
-					&& !param.typ.has_flag(.generic) && expected_type != typ {
-					styp := c.table.type_to_str(typ)
-					elem_styp := c.table.type_to_str(expected_type)
+				is_decompose := arg.expr is ast.ArrayDecompose
+				styp := c.table.type_to_str(typ)
+				elem_styp := c.table.type_to_str(expected_type)
+				is_single_array_arg := node.args.len == 1 && i == 0
+				sum_type_needs_spread := c.table.sym(expected_type).kind == .sum_type
+					&& is_single_array_arg
+				if c.table.sym(typ).kind == .array && !is_decompose
+					&& (c.table.sym(expected_type).kind !in [.sum_type, .interface]
+					|| sum_type_needs_spread) && !param.typ.has_flag(.generic)
+					&& expected_type != typ {
 					c.error('to pass `${arg.expr}` (${styp}) to `${method.name}` (which accepts type `...${elem_styp}`), use `...${arg.expr}`',
 						node.pos)
-				} else if arg.expr is ast.ArrayDecompose
-					&& c.table.sym(expected_type).kind == .sum_type
-					&& expected_type.idx() != typ.idx() {
+				} else if is_decompose && !param.typ.has_flag(.generic)
+					&& c.table.sym(expected_type).kind != .interface
+					&& expected_type.idx() != typ.idx() && typ != ast.void_type {
 					expected_type_str := c.table.type_to_str(expected_type)
-					got_type_str := c.table.type_to_str(typ)
-					c.error('cannot use `...${got_type_str}` as `...${expected_type_str}` in argument ${
-						i + 1} to `${method_name}`', arg.pos)
+					c.error('cannot use `...${styp}` as `...${expected_type_str}` in argument ${i +
+						1} to `${method.name}`', arg.pos)
 				}
 			}
 		} else {

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1528,40 +1528,9 @@ fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool) ast.
 				c.expr(mut call_arg.expr)
 			}
 			if i == args_len - 1 {
-				kind := c.table.sym(typ).kind
-				is_decompose := call_arg.expr is ast.ArrayDecompose
-				mut allow_variadic_pass := false
-				if call_arg.expr is ast.Ident && !func.is_method {
-					ident := call_arg.expr as ast.Ident
-					if ident.obj is ast.Var {
-						var_obj := ident.obj as ast.Var
-						if var_obj.is_arg && c.table.cur_fn != unsafe { nil }
-							&& c.table.cur_fn.is_variadic && c.table.cur_fn.params.len > 0
-							&& ident.name == c.table.cur_fn.params.last().name && func.is_variadic {
-							cur_fn_param_sym := c.table.sym(c.table.cur_fn.params.last().typ)
-							if cur_fn_param_sym.info is ast.Array {
-								allow_variadic_pass = cur_fn_param_sym.info.elem_type == expected_type
-							}
-						}
-					}
-				}
-				styp := c.table.type_to_str(typ)
-				elem_styp := c.table.type_to_str(expected_type)
-				is_single_array_arg := args_len == 1 && i == 0
-				sum_type_needs_spread := c.table.sym(expected_type).kind == .sum_type
-					&& is_single_array_arg
-				if kind == .array && !is_decompose && !allow_variadic_pass
-					&& (c.table.sym(expected_type).kind !in [.sum_type, .interface]
-					|| sum_type_needs_spread) && !param.typ.has_flag(.generic)
-					&& expected_type != typ {
-					c.error('to pass `${call_arg.expr}` (${styp}) to `${func.name}` (which accepts type `...${elem_styp}`), use `...${call_arg.expr}`',
-						node.pos)
-				} else if is_decompose && !param.typ.has_flag(.generic)
-					&& c.table.sym(expected_type).kind != .interface
-					&& expected_type.idx() != typ.idx() && typ != ast.void_type {
-					c.error('cannot use `...${styp}` as `...${elem_styp}` in argument ${i + 1} to `${fn_name}`',
-						call_arg.pos)
-				}
+				c.check_variadic_arg(call_arg.expr, typ, expected_type, param.typ, i + 1,
+					func.name, func.is_method, func.is_variadic, args_len == 1 && i == 0,
+					node.pos, call_arg.pos)
 			}
 		} else {
 			c.expected_type = param.typ
@@ -2573,25 +2542,9 @@ fn (mut c Checker) method_call(mut node ast.CallExpr, mut continue_check &bool) 
 			}
 			typ := c.expr(mut arg.expr)
 			if i == node.args.len - 1 {
-				is_decompose := arg.expr is ast.ArrayDecompose
-				styp := c.table.type_to_str(typ)
-				elem_styp := c.table.type_to_str(expected_type)
-				is_single_array_arg := node.args.len == 1 && i == 0
-				sum_type_needs_spread := c.table.sym(expected_type).kind == .sum_type
-					&& is_single_array_arg
-				if c.table.sym(typ).kind == .array && !is_decompose
-					&& (c.table.sym(expected_type).kind !in [.sum_type, .interface]
-					|| sum_type_needs_spread) && !param.typ.has_flag(.generic)
-					&& expected_type != typ {
-					c.error('to pass `${arg.expr}` (${styp}) to `${method.name}` (which accepts type `...${elem_styp}`), use `...${arg.expr}`',
-						node.pos)
-				} else if is_decompose && !param.typ.has_flag(.generic)
-					&& c.table.sym(expected_type).kind != .interface
-					&& expected_type.idx() != typ.idx() && typ != ast.void_type {
-					expected_type_str := c.table.type_to_str(expected_type)
-					c.error('cannot use `...${styp}` as `...${expected_type_str}` in argument ${i +
-						1} to `${method.name}`', arg.pos)
-				}
+				c.check_variadic_arg(arg.expr, typ, expected_type, param.typ, i + 1, method.name,
+					true, method.is_variadic, node.args.len == 1 && i == 0, node.pos,
+					arg.pos)
 			}
 		} else {
 			c.expected_type = param.typ
@@ -4014,4 +3967,40 @@ fn (mut c Checker) has_veb_context(typ ast.Type) bool {
 		}
 	}
 	return false
+}
+
+fn (mut c Checker) check_variadic_arg(arg_expr ast.Expr, typ ast.Type, expected_type ast.Type, param_typ ast.Type,
+	arg_num int, fn_name string, is_method bool, fn_is_variadic bool, is_single_array_arg bool,
+	call_pos token.Pos, arg_pos token.Pos) {
+	kind := c.table.sym(typ).kind
+	is_decompose := arg_expr is ast.ArrayDecompose
+	mut allow_variadic_pass := false
+	if arg_expr is ast.Ident && !is_method {
+		ident := arg_expr as ast.Ident
+		if ident.obj is ast.Var {
+			var_obj := ident.obj as ast.Var
+			if var_obj.is_arg && c.table.cur_fn != unsafe { nil } && c.table.cur_fn.is_variadic
+				&& c.table.cur_fn.params.len > 0 && ident.name == c.table.cur_fn.params.last().name
+				&& fn_is_variadic {
+				cur_fn_param_sym := c.table.sym(c.table.cur_fn.params.last().typ)
+				if cur_fn_param_sym.info is ast.Array {
+					allow_variadic_pass = cur_fn_param_sym.info.elem_type == expected_type
+				}
+			}
+		}
+	}
+	styp := c.table.type_to_str(typ)
+	elem_styp := c.table.type_to_str(expected_type)
+	expected_kind := c.table.sym(expected_type).kind
+	sum_type_needs_spread := expected_kind == .sum_type && is_single_array_arg
+	if kind == .array && !is_decompose && !allow_variadic_pass
+		&& (expected_kind !in [.sum_type, .interface] || sum_type_needs_spread)
+		&& !param_typ.has_flag(.generic) && expected_type != typ {
+		c.error('to pass `${arg_expr}` (${styp}) to `${fn_name}` (which accepts type `...${elem_styp}`), use `...${arg_expr}`',
+			call_pos)
+	} else if is_decompose && !param_typ.has_flag(.generic) && expected_kind != .interface
+		&& expected_type.idx() != typ.idx() && typ != ast.void_type {
+		c.error('cannot use `...${styp}` as `...${elem_styp}` in argument ${arg_num} to `${fn_name}`',
+			arg_pos)
+	}
 }

--- a/vlib/v/checker/tests/sumtype_array_to_variadic.out
+++ b/vlib/v/checker/tests/sumtype_array_to_variadic.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/sumtype_array_to_variadic.vv:32:9: error: to pass `kv_pairs` ([]Value) to `Map__static__create` (which accepts type `...Value`), use `...kv_pairs`
+   30 |     }
+   31 |     assert kv_pairs.len % 2 == 0
+   32 |     return Map.create(kv_pairs)
+      |            ~~~~~~~~~~~~~~~~~~~~
+   33 | }
+   34 |

--- a/vlib/v/checker/tests/sumtype_array_to_variadic.vv
+++ b/vlib/v/checker/tests/sumtype_array_to_variadic.vv
@@ -1,0 +1,38 @@
+type Nil = u8
+
+const c_nil = Nil(`\0`)
+
+pub type Value = Nil | int | string | []Value | Map
+
+pub fn (v Value) string() string {
+	return '${v.str()}'
+}
+
+pub struct Map {
+mut:
+	m map[string]Value
+}
+
+pub fn Map.create(values ...Value) Map {
+	assert values.len > 1
+	assert values.len % 2 == 0
+	mut m := Map{}
+	for i := 0; i < values.len; i += 2 {
+		m.m[values[i].string()] = values[i + 1]
+	}
+	return m
+}
+
+pub fn build_map(index_from int, index_to int) Map {
+	mut kv_pairs := []Value{len: index_to - index_from, init: Value(c_nil)}
+	for i := index_from; i < index_to; i++ {
+		kv_pairs[i - index_from] = Value(i)
+	}
+	assert kv_pairs.len % 2 == 0
+	return Map.create(kv_pairs)
+}
+
+fn main() {
+	m := build_map(2, 4)
+	_ = m
+}

--- a/vlib/v/tests/fns/call_args_variadic_sumtype_test.v
+++ b/vlib/v/tests/fns/call_args_variadic_sumtype_test.v
@@ -20,5 +20,5 @@ fn test_main() {
 	}
 	mut animals := []Animal{}
 	animals << cat
-	print_names(animals)
+	print_names(...animals)
 }

--- a/vlib/v/tests/fns/variadic_sumtype_test.v
+++ b/vlib/v/tests/fns/variadic_sumtype_test.v
@@ -14,6 +14,11 @@ fn div(params ...Either) int {
 	return tag(params)
 }
 
+fn div2(params ...Either) int {
+	return tag(...params)
+}
+
 fn test_main() {
 	assert dump(div('foo', 1, 2)) == 3
+	assert dump(div2(6, -2, 'hello', 10)) == 14
 }

